File: components/digid_ns/src/main/java/nl/logius/digid/ns/model/Registration.java

/*
  Deze broncode is openbaar gemaakt vanwege een Woo-verzoek zodat deze
  gericht is op transparantie en niet op hergebruik. Hergebruik van 
  de broncode is toegestaan onder de EUPL licentie, met uitzondering 
  van broncode waarvoor een andere licentie is aangegeven.
  
  Het archief waar dit bestand deel van uitmaakt is te vinden op:
    https://github.com/MinBZK/woo-besluit-broncode-digid
  
  Eventuele kwetsbaarheden kunnen worden gemeld bij het NCSC via:
    https://www.ncsc.nl/contact/kwetsbaarheid-melden
  onder vermelding van "Logius, openbaar gemaakte broncode DigiD" 
  
  Voor overige vragen over dit Woo-besluit kunt u mailen met open@logius.nl
  
  This code has been disclosed in response to a request under the Dutch
  Open Government Act ("Wet open Overheid"). This implies that publication 
  is primarily driven by the need for transparence, not re-use.
  Re-use is permitted under the EUPL-license, with the exception 
  of source files that contain a different license.
  
  The archive that this file originates from can be found at:
    https://github.com/MinBZK/woo-besluit-broncode-digid
  
  Security vulnerabilities may be responsibly disclosed via the Dutch NCSC:
    https://www.ncsc.nl/contact/kwetsbaarheid-melden
  using the reference "Logius, publicly disclosed source code DigiD" 
  
  Other questions regarding this Open Goverment Act decision may be
  directed via email to open@logius.nl
*/

package nl.logius.digid.ns.model;

import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import javax.persistence.*;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;

@Entity
@Table(name = "notification_registrations")
public class Registration {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String appId;
    private String notificationId;
    private Long accountId;
    private String deviceName;
    private boolean receiveNotifications;
    private boolean registrationSucceeded;
    private int osType;
    private String osVersion;

    public Registration() {}

    public Registration(String appId, String notificationId,  Long accountId, String deviceName, boolean receiveNotifications, int osType, String osVersion){
        this.notificationId = notificationId;
        this.appId = appId;
        this.accountId = accountId;
        this.deviceName = deviceName;
        this.receiveNotifications = receiveNotifications;
        this.osType = osType;
        this.osVersion = osVersion;
    }

    public ZonedDateTime timestampToZonedDateTime(String text) {
        return ZonedDateTime.of(LocalDateTime.parse(text.substring(0, 19)), ZoneOffset.UTC);
    }

    @Column(name = "created_at", updatable=false)
    @CreationTimestamp
    private ZonedDateTime createdAt;
    @Column(name = "updated_at")
    @UpdateTimestamp
    private ZonedDateTime updatedAt;


    public ZonedDateTime getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(ZonedDateTime createdAt) {
        this.createdAt = createdAt;
    }

    public ZonedDateTime getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(ZonedDateTime updatedAt) {
        this.updatedAt = updatedAt;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getNotificationId() {
        return notificationId;
    }

    public void setNotificationId(String notificationId) {
        this.notificationId = notificationId;
    }

    public String getAppId() {
        return appId;
    }

    public void setAppId(String appId) {
        this.appId = appId;
    }

    public Long getAccountId() {
        return accountId;
    }

    public void setAccountId(Long accountId) {
        this.accountId = accountId;
    }

    public String getDeviceName() {
        return deviceName;
    }

    public void setDeviceName(String deviceName) {
        this.deviceName = deviceName;
    }

    public boolean isReceiveNotifications() {
        return receiveNotifications;
    }

    public void setReceiveNotifications(boolean receiveNotifications) {
        this.receiveNotifications = receiveNotifications;
    }

    public boolean isRegistrationSucceeded() {
        return registrationSucceeded;
    }

    public void setRegistrationSucceeded(boolean registrationSucceeded) {
        this.registrationSucceeded = registrationSucceeded;
    }

    public int getOsType() {
        return osType;
    }

    public void setOsType(int osType) {
        this.osType = osType;
    }

    public String getOsVersion() {
        return osVersion;
    }

    public void setOsVersion(String osVersion) {
        this.osVersion = osVersion;
    }
}

File: components/digid_admin/app/models/registration.rb

# Deze broncode is openbaar gemaakt vanwege een Woo-verzoek zodat deze
# gericht is op transparantie en niet op hergebruik. Hergebruik van 
# de broncode is toegestaan onder de EUPL licentie, met uitzondering 
# van broncode waarvoor een andere licentie is aangegeven.
# 
# Het archief waar dit bestand deel van uitmaakt is te vinden op:
#   https://github.com/MinBZK/woo-besluit-broncode-digid
# 
# Eventuele kwetsbaarheden kunnen worden gemeld bij het NCSC via:
#   https://www.ncsc.nl/contact/kwetsbaarheid-melden
# onder vermelding van "Logius, openbaar gemaakte broncode DigiD" 
# 
# Voor overige vragen over dit Woo-besluit kunt u mailen met open@logius.nl
# 
# This code has been disclosed in response to a request under the Dutch
# Open Government Act ("Wet open Overheid"). This implies that publication 
# is primarily driven by the need for transparence, not re-use.
# Re-use is permitted under the EUPL-license, with the exception 
# of source files that contain a different license.
# 
# The archive that this file originates from can be found at:
#   https://github.com/MinBZK/woo-besluit-broncode-digid
# 
# Security vulnerabilities may be responsibly disclosed via the Dutch NCSC:
#   https://www.ncsc.nl/contact/kwetsbaarheid-melden
# using the reference "Logius, publicly disclosed source code DigiD" 
# 
# Other questions regarding this Open Goverment Act decision may be
# directed via email to open@logius.nl

class Registration < AccountBase
  has_many :activation_letters, autosave: true, dependent: :destroy
  belongs_to :nationality

  alias_attribute :name, :burgerservicenummer

  module Status
    ABORTED   = 'aborted'.freeze # afgebroken
    COMPLETED = 'completed'.freeze
    INITIAL   = 'initial'.freeze # initieel
    REQUESTED = 'requested'.freeze # aangevraagd
  end

  include Stateful

  def self.month_count(bsn)
    bsn_is(bsn).with_valid_recovery_code.in_last_month.count
  end

  def self.bsn_is(bsn)
    where burgerservicenummer: bsn
  end

  def self.with_valid_recovery_code
    where gba_status: 'valid_recovery'
  end

  def self.in_last_month
    where 'created_at > ?', Time.now - 1.month
  end

  def self.too_soon
    snelheid_herstelcode = ::Configuration.get_int('snelheid_herstelcode')
    where 'created_at > ?', Time.now - snelheid_herstelcode.hours
  end

  def self.postcodes(postcode)
    where postcode: postcode
  end

  def self.valid
    where gba_status: 'valid'
  end

  def self.find_bsns_from_postcodes(content)
    bsns = []
    content.each do |postcode|
      bsns << Registration.valid.postcodes(postcode.gsub(/\s+/, '')).pluck(:burgerservicenummer)
    end
    bsns.flatten.uniq
  end
end
